File: server/internal/handler/calendar.go
package handler

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/unicomhub/uniwork/server/internal/handler/dto/sdo"
	"github.com/unicomhub/uniwork/server/internal/middleware"
	"github.com/unicomhub/uniwork/server/internal/service"
	"github.com/unicomhub/uniwork/server/internal/util"
)

func (h *handlers) listCalendarEvents(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	fromRaw := strings.TrimSpace(q.Get("from"))
	toRaw := strings.TrimSpace(q.Get("to"))
	if fromRaw == "" || toRaw == "" {
		respondError(w, http.StatusBadRequest, "invalid_request", "from and to are required (YYYY-MM-DD)")
		return
	}
	fromDate, err := util.ParseCalendarDate(fromRaw)
	if err != nil {
		respondError(w, http.StatusBadRequest, "invalid_request", "from must be YYYY-MM-DD")
		return
	}
	toDate, err := util.ParseCalendarDate(toRaw)
	if err != nil {
		respondError(w, http.StatusBadRequest, "invalid_request", "to must be YYYY-MM-DD")
		return
	}
	mine := false
	if raw := strings.TrimSpace(q.Get("mine")); raw != "" {
		mine, err = strconv.ParseBool(raw)
		if err != nil {
			respondError(w, http.StatusBadRequest, "invalid_request", "mine must be true or false")
			return
		}
	}

	events, err := h.Calendar.ListEvents(
		r.Context(),
		chi.URLParam(r, "workspaceID"),
		middleware.UserID(r.Context()),
		fromDate.Time,
		toDate.Time,
		mine,
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	out := sdo.CalendarEventListSDO{Events: make([]sdo.CalendarEventSDO, 0, len(events))}
	for _, e := range events {
		out.Events = append(out.Events, toCalendarEventSDO(e))
	}
	respondJSON(w, http.StatusOK, out)
}

func (h *handlers) workspaceCalendar(w http.ResponseWriter, r *http.Request) {
	wsID := chi.URLParam(r, "workspaceID")
	userID := middleware.UserID(r.Context())
	q := r.URL.Query()
	fromRaw := strings.TrimSpace(q.Get("from"))
	toRaw := strings.TrimSpace(q.Get("to"))

	var fromDay, toDay time.Time
	switch {
	case fromRaw == "" && toRaw == "":
		today := time.Now().UTC().Truncate(24 * time.Hour)
		fromDay = today.AddDate(0, 0, -30)
		toDay = today.AddDate(0, 0, 90)
	case fromRaw != "" && toRaw != "":
		fromDate, err := util.ParseCalendarDate(fromRaw)
		if err != nil {
			respondError(w, http.StatusBadRequest, "invalid_request", "from must be YYYY-MM-DD")
			return
		}
		toDate, err := util.ParseCalendarDate(toRaw)
		if err != nil {
			respondError(w, http.StatusBadRequest, "invalid_request", "to must be YYYY-MM-DD")
			return
		}
		fromDay = fromDate.Time
		toDay = toDate.Time
	default:
		respondError(w, http.StatusBadRequest, "invalid_request", "from and to must both be set or both omitted")
		return
	}

	ics, err := h.Calendar.WorkspaceICS(r.Context(), wsID, userID, fromDay, toDay)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	w.Header().Set("Content-Type", "text/calendar; charset=utf-8")
	w.Header().Set("Content-Disposition", `attachment; filename="uniwork-calendar.ics"`)
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(ics)
}

func (h *handlers) listCalendarSidebar(w http.ResponseWriter, r *http.Request) {
	sidebar, err := h.Calendar.ListSidebar(
		r.Context(),
		chi.URLParam(r, "workspaceID"),
		middleware.UserID(r.Context()),
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, toCalendarSidebarSDO(sidebar))
}

func toCalendarSidebarSDO(s service.CalendarSidebar) sdo.CalendarSidebarSDO {
	out := sdo.CalendarSidebarSDO{
		Priorities:   make([]sdo.CalendarSidebarTaskSDO, 0, len(s.Priorities)),
		MeetWith:     make([]sdo.CalendarSidebarMeetingSDO, 0, len(s.MeetWith)),
		Assigned:     make([]sdo.CalendarSidebarTaskSDO, 0, len(s.Assigned)),
		TodayOverdue: make([]sdo.CalendarSidebarTaskSDO, 0, len(s.TodayOverdue)),
		Backlog:      make([]sdo.CalendarSidebarTaskSDO, 0, len(s.Backlog)),
	}
	for _, t := range s.Priorities {
		out.Priorities = append(out.Priorities, toCalendarSidebarTaskSDO(t))
	}
	for _, m := range s.MeetWith {
		out.MeetWith = append(out.MeetWith, sdo.CalendarSidebarMeetingSDO{
			ID: m.ID, Title: m.Title, StartsAt: m.StartsAt, EndsAt: m.EndsAt,
		})
	}
	for _, t := range s.Assigned {
		out.Assigned = append(out.Assigned, toCalendarSidebarTaskSDO(t))
	}
	for _, t := range s.TodayOverdue {
		out.TodayOverdue = append(out.TodayOverdue, toCalendarSidebarTaskSDO(t))
	}
	for _, t := range s.Backlog {
		out.Backlog = append(out.Backlog, toCalendarSidebarTaskSDO(t))
	}
	return out
}

func toCalendarSidebarTaskSDO(t service.CalendarSidebarTask) sdo.CalendarSidebarTaskSDO {
	return sdo.CalendarSidebarTaskSDO{
		ID: t.ID, Title: t.Title, Status: t.Status, Priority: t.Priority, DueDate: t.DueDate,
	}
}

func toCalendarEventSDO(e service.CalendarEvent) sdo.CalendarEventSDO {
	return sdo.CalendarEventSDO{
		ID:        e.ID,
		Kind:      e.Kind,
		EntityID:  e.EntityID,
		Title:     e.Title,
		Start:     e.Start,
		End:       e.End,
		AllDay:    e.AllDay,
		Status:    e.Status,
		Priority:  e.Priority,
		ProjectID: e.ProjectID,
	}
}

File: packages/core/api/endpoints/calendar.ts
import { z } from "zod";
import type { CalendarEvent, CalendarSidebar } from "../../calendar/types";
import { ApiError, request, requestText } from "../http";
import { parseWithFallback } from "../schema";

const enc = encodeURIComponent;

const CalendarEventItemSchema = z
  .object({
    id: z.string(),
    kind: z.string(),
    entity_id: z.string(),
    title: z.string(),
    start: z.string(),
    end: z.string().optional().nullable(),
    all_day: z.boolean(),
    status: z.string().optional().nullable(),
    priority: z.string().optional().nullable(),
    project_id: z.string().optional().nullable(),
  })
  .transform(
    (e): CalendarEvent => ({
      id: e.id,
      kind: e.kind as CalendarEvent["kind"],
      entityId: e.entity_id,
      title: e.title,
      start: e.start,
      end: e.end ?? undefined,
      allDay: e.all_day,
      status: e.status ?? undefined,
      priority: e.priority ?? undefined,
      projectId: e.project_id ?? undefined,
    }),
  );

const CalendarEventListSchema = z
  .object({
    events: z.array(CalendarEventItemSchema),
  })
  .transform((v) => v.events);

const EMPTY_EVENTS: CalendarEvent[] = [];

const CalendarSidebarTaskSchema = z
  .object({
    id: z.string(),
    title: z.string(),
    status: z.string(),
    priority: z.string().optional().nullable(),
    due_date: z.string().optional().nullable(),
  })
  .transform((t) => ({
    id: t.id,
    title: t.title,
    status: t.status,
    priority: t.priority ?? undefined,
    dueDate: t.due_date ?? undefined,
  }));

const CalendarSidebarMeetingSchema = z
  .object({
    id: z.string(),
    title: z.string(),
    starts_at: z.string(),
    ends_at: z.string(),
  })
  .transform((m) => ({
    id: m.id,
    title: m.title,
    startsAt: m.starts_at,
    endsAt: m.ends_at,
  }));

const CalendarSidebarSchema = z
  .object({
    priorities: z.array(CalendarSidebarTaskSchema).optional(),
    meet_with: z.array(CalendarSidebarMeetingSchema).optional(),
    assigned: z.array(CalendarSidebarTaskSchema).optional(),
    today_overdue: z.array(CalendarSidebarTaskSchema).optional(),
    backlog: z.array(CalendarSidebarTaskSchema).optional(),
  })
  .transform(
    (v): CalendarSidebar => ({
      priorities: v.priorities ?? [],
      meetWith: v.meet_with ?? [],
      assigned: v.assigned ?? [],
      todayOverdue: v.today_overdue ?? [],
      backlog: v.backlog ?? [],
    }),
  );

const EMPTY_SIDEBAR: CalendarSidebar = {
  priorities: [],
  meetWith: [],
  assigned: [],
  todayOverdue: [],
  backlog: [],
};

export async function listCalendarEvents(
  workspaceId: string,
  params: { from: string; to: string; mine?: boolean },
): Promise<CalendarEvent[]> {
  const q = new URLSearchParams({ from: params.from, to: params.to });
  if (params.mine) q.set("mine", "true");
  const raw = await request(`/api/v1/workspaces/${enc(workspaceId)}/calendar/events?${q}`);
  return parseWithFallback<CalendarEvent[]>(raw, CalendarEventListSchema, EMPTY_EVENTS, {
    endpoint: "GET /api/v1/workspaces/{ws}/calendar/events",
  });
}

export async function getCalendarSidebar(workspaceId: string): Promise<CalendarSidebar> {
  const raw = await request(`/api/v1/workspaces/${enc(workspaceId)}/calendar/sidebar`);
  return parseWithFallback<CalendarSidebar>(raw, CalendarSidebarSchema, EMPTY_SIDEBAR, {
    endpoint: "GET /api/v1/workspaces/{ws}/calendar/sidebar",
  });
}

const WORKSPACE_ICS_FILENAME = "uniwork-calendar.ics";

/** Fetches workspace iCalendar text; caller saves it as a download. */
export async function fetchWorkspaceCalendarIcs(
  workspaceId: string,
  opts?: { from?: string; to?: string },
): Promise<string> {
  const q = new URLSearchParams();
  if (opts?.from) q.set("from", opts.from);
  if (opts?.to) q.set("to", opts.to);
  const query = q.toString();
  const path = `/api/v1/workspaces/${enc(workspaceId)}/calendar.ics${query ? `?${query}` : ""}`;
  const text = await requestText(path);
  if (!text.trim() || !text.includes("BEGIN:VCALENDAR")) {
    throw new ApiError("Invalid calendar export", "internal", 502);
  }
  return text;
}

export { WORKSPACE_ICS_FILENAME };
